src/backend/distributed/planner/multi_join_order.c

```diff
@@ -398,6 +399,42 @@ ExtractLeftMostRangeTableIndex(Node *node, int *rangeTableIndex)
 
 
 /*
+ * GetRightRangeTableRef returns the range table reference from the right
+ * branch of a join expression. Right child of the join expresion is expected
+ * to be RangeTblRef*. When postgresql standard planner flattens a multi level
+ * query into a single one it tweaks join expr tree and puts FromExpr * into
+ * rarg field. Depending on the depth of the original query there could be
+ * nested FromExpr * inside the other one as the only element in fromList.
+ */
+static RangeTblRef *
+GetRightRangeTableRef(Node *joinExprRightArg)
+{
+	RangeTblRef *rangeTableRef  = NULL;
+	if (IsA(joinExprRightArg, RangeTblRef))
+	{
+		rangeTableRef = (RangeTblRef *) joinExprRightArg;
+	}
+	else if (IsA(joinExprRightArg, FromExpr))
+	{
+		FromExpr *fromExpr = (FromExpr *) joinExprRightArg;
+		Node *firstNode = NULL;
+
+		Assert(list_length(fromExpr->fromlist) == 1);
```

Comment:
I will add that to HasComplexJoinOrder where we report error depending on the result of the function.